File: 剑指offer代码/24_verifySequenceBST.py
'''
@Author: Darcy
@Date: May, 19, 2017
@Topic: Verify if the sequence is the postorder of binary search tree
'''
def verifySequenceBST(sequence):
	if sequence == None or sequence == []:
		return False
	root = sequence[-1]
	index = 0
	for i in range(len(sequence) - 1):
		if sequence[i] > root:
			index = i
			break
	for j in range(index, len(sequence) - 1):
		if sequence[j] < root:
			return False
	if index > 0:
		left = verifySequenceBST(sequence[:index])
	if index < len(sequence) - 1:
		right = verifySequenceBST(sequence[index:len(sequence) - 1])
	return right and left

if __name__ == "__main__":
	arr = [int(val) for val in raw_input().strip().split()]
	print verifySequenceBST(arr)

File: 剑指offer代码/24_verifySequenceBST.cpp
bool verifybst(int sequence[], int length)
{
	if(sequence == NULL || length == 0)
		return false;
	int root = sequence[length - 1];
	int i = 0;
	for(; i < length; ++i)
	{
		if(sequence[i] > root)
			break;
	}
	int j = i;
	for(; j < length - 1; ++j)
	{
		if(sequence[j] < root)
			return false;
	}
	bool left = true;
	if(i > 0)
		left = verifybst(sequence, i);
	bool right = true;
	if(i < length - 1)
		right = verifybst(sequence + i, length - i - 1);
	return (left && right);
}